pkg/operator/injection/injection.go

```diff
@@ -76,36 +61,49 @@ func GetControllerName(ctx context.Context) string {
 	return name.(string)
 }
 
-// WithSettingsOrDie injects the settings into the context for all configMaps passed through the registrations
+func WithOptions(ctx context.Context, injectables ...options.Injectable) context.Context {
+	for _, opts := range injectables {
+		ctx = lo.Must(opts.Inject(ctx))
+	}
+	return ctx
+}
+
+// WithSettings injects the settings into the context for all configMaps passed through the registrations
 // NOTE: Settings are resolved statically into the global context.Context at startup. This was changed from updating them
 // dynamically at runtime due to the necessity of having to build logic around re-queueing to ensure that settings are
 // properly reloaded for things like feature gates
-func WithSettingsOrDie(ctx context.Context, kubernetesInterface kubernetes.Interface, settings ...settings.Injectable) context.Context {
+func WithSettings(ctx context.Context, kubernetesInterface kubernetes.Interface, options []options.Injectable, injectables ...settings.Injectable) context.Context {
 	cancelCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	factory := informers.NewSharedInformerFactoryWithOptions(kubernetesInterface, time.Second*30, informers.WithNamespace(system.Namespace()))
 	informer := factory.Core().V1().ConfigMaps().Informer()
 	factory.Start(cancelCtx.Done())
+	factory.WaitForCacheSync(cancelCtx.Done())
 
-	for _, setting := range settings {
-		cm := lo.Must(waitForConfigMap(ctx, setting.ConfigMap(), informer))
+	for _, setting := range injectables {
+		cm := lo.Must(getConfigMap(setting.ConfigMap(), informer))
+		if cm == nil {
+			continue
+		}
 		ctx = lo.Must(setting.Inject(ctx, cm))
 	}
+
+	for _, opts := range options {
+		opts.MergeSettings(ctx, injectables...)
+	}
+
 	return ctx
 }
 
-// waitForConfigMap waits until all registered configMaps in the settingsStore are created
-func waitForConfigMap(ctx context.Context, name string, informer cache.SharedIndexInformer) (*v1.ConfigMap, error) {
-	for {
-		configMap, exists, err := informer.GetStore().GetByKey(types.NamespacedName{Namespace: system.Namespace(), Name: name}.String())
-		if configMap != nil && exists && err == nil {
-			return configMap.(*v1.ConfigMap), nil
-		}
-		select {
-		case <-ctx.Done():
-			return nil, fmt.Errorf("context canceled")
-		case <-time.After(time.Millisecond * 500):
-		}
+// getConfigMap waits until all registered configMaps in the settingsStore are created
+func getConfigMap(name string, informer cache.SharedIndexInformer) (*v1.ConfigMap, error) {
```

Comment:
We discussed this but I don't think that we need to change this logic at all. We can just create a context that has a timeout and then based on if we don't resolve the ConfigMap, we can continue forward. Specifically, we should check if we are getting NotFound errors and not just transient errors from the kube client API calls. See: https://github.com/aws/karpenter-core/pull/544/files#diff-49fe45488dc421bbc55f13e760b3b06b8755016ae17c70915eb0b44752b5f476R96